File: software/host/src/core/DataProcessor.cpp
/**
 * HeteroLink Host - 数据处理器实现
 */

#include "core/DataProcessor.h"
#include "utils/Logger.h"
#include <QFile>
#include <QTextStream>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonArray>
#include <cmath>

namespace HeteroLink {

DataProcessor::DataProcessor(QObject *parent)
    : QObject(parent)
{
}

DataProcessor::~DataProcessor()
{
    qDeleteAll(deviceData_);
}

void DataProcessor::setBufferSize(const QString& deviceId, int size)
{
    auto it = deviceData_.find(deviceId);
    if (it == deviceData_.end()) {
        deviceData_[deviceId] = new DeviceData(size);
    } else {
        (*it)->maxBufferSize = size;
        // 修剪缓冲区
        while ((*it)->buffer.size() > size) {
            (*it)->buffer.dequeue();
        }
    }
    LOG_INFO("Buffer size set for device " + deviceId + ": " + QString::number(size));
}

void DataProcessor::addData(const QString& deviceId, const TelemetryData& data)
{
    auto it = deviceData_.find(deviceId);
    if (it == deviceData_.end()) {
        deviceData_[deviceId] = new DeviceData(defaultBufferSize_);
        it = deviceData_.find(deviceId);
    }
    
    (*it)->buffer.enqueue(data);
    
    // 保持缓冲区大小限制
    while ((*it)->buffer.size() > (*it)->maxBufferSize) {
        (*it)->buffer.dequeue();
    }
    
    (*it)->lastUpdate = QDateTime::currentMSecsSinceEpoch();
    
    updateStats(deviceId, data);
    
    emit dataUpdated(deviceId);
}

QVector<TelemetryData> DataProcessor::getData(const QString& deviceId) const
{
    auto it = deviceData_.find(deviceId);
    if (it == deviceData_.end()) {
        return QVector<TelemetryData>();
    }
    
    QVector<TelemetryData> result;
    const auto& buffer = (*it)->buffer;
    result.reserve(buffer.size());
    for (const auto& item : buffer) {
        result.append(item);
    }
    return result;
}

QVector<TelemetryData> DataProcessor::getLatestData(const QString& deviceId, int count) const
{
    auto it = deviceData_.find(deviceId);
    if (it == deviceData_.end()) {
        return QVector<TelemetryData>();
    }
    
    const auto& buffer = (*it)->buffer;
    int start = qMax(0, buffer.size() - count);
    QVector<TelemetryData> result;
    result.reserve(buffer.size() - start);
    
    auto bit = buffer.begin();
    std::advance(bit, start);
    for (; bit != buffer.end(); ++bit) {
        result.append(*bit);
    }
    return result;
}

QMap<int, ChannelStats> DataProcessor::getStats(const QString& deviceId) const
{
    auto it = deviceData_.find(deviceId);
    if (it == deviceData_.end()) {
        return QMap<int, ChannelStats>();
    }
    
    return (*it)->channelStats;
}

void DataProcessor::clearData(const QString& deviceId)
{
    auto it = deviceData_.find(deviceId);
    if (it != deviceData_.end()) {
        (*it)->buffer.clear();
        (*it)->channelStats.clear();
        emit dataUpdated(deviceId);
        emit statsUpdated(deviceId);
    }
}

void DataProcessor::clearAll()
{
    deviceData_.clear();
    LOG_INFO("All data cleared");
}

bool DataProcessor::exportToCsv(const QString& deviceId, const QString& filePath) const
{
    auto it = deviceData_.find(deviceId);
    if (it == deviceData_.end()) {
        LOG_ERROR("No data for device: " + deviceId);
        return false;
    }
    
    QFile file(filePath);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
        LOG_ERROR("Failed to open file: " + filePath);
        return false;
    }
    
    QTextStream out(&file);
    
    // 写入表头
    const auto& buffer = (*it)->buffer;
    if (buffer.isEmpty()) {
        return true;
    }
    
    // 转换为 QVector 以便索引访问
    QVector<TelemetryData> data;
    data.reserve(buffer.size());
    for (const auto& item : buffer) {
        data.append(item);
    }
    
    int channelCount = data[0].channels.size();
    out << "timestamp";
    for (int i = 0; i < channelCount; ++i) {
        out << ",channel" << i;
    }
    out << "\n";
    
    // 写入数据
    for (int i = 0; i < data.size(); ++i) {
        out << data[i].timestamp;
        for (int ch = 0; ch < data[i].channels.size(); ++ch) {
            out << "," << data[i].channels[ch];
        }
        out << "\n";
    }
    
    file.close();
    LOG_INFO("Data exported to CSV: " + filePath);
    return true;
}

bool DataProcessor::exportToJson(const QString& deviceId, const QString& filePath) const
{
    auto it = deviceData_.find(deviceId);
    if (it == deviceData_.end()) {
        LOG_ERROR("No data for device: " + deviceId);
        return false;
    }
    
    const auto& buffer = (*it)->buffer;
    QJsonArray jsonArray;
    
    for (const auto& item : buffer) {
        QJsonObject obj;
        obj["timestamp"] = static_cast<qint64>(item.timestamp);
        
        QJsonArray channels;
        for (const auto& ch : item.channels) {
            channels.append(ch);
        }
        obj["channels"] = channels;
        
        jsonArray.append(obj);
    }
    
    QJsonDocument doc(jsonArray);
    
    QFile file(filePath);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
        LOG_ERROR("Failed to open file: " + filePath);
        return false;
    }
    
    file.write(doc.toJson());
    file.close();
    LOG_INFO("Data exported to JSON: " + filePath);
    return true;
}

void DataProcessor::updateStats(const QString& deviceId, const TelemetryData& data)
{
    auto it = deviceData_.find(deviceId);
    if (it == deviceData_.end()) {
        return;
    }
    
    auto& devData = (*it);
    
    for (int i = 0; i < data.channels.size(); ++i) {
        float value = data.channels[i];
        
        if (!devData->channelStats.contains(i)) {
            devData->channelStats[i] = ChannelStats();
        }
        
        auto& stats = devData->channelStats[i];
        
        // 更新最小/最大值
        if (stats.sampleCount == 0 || value < stats.min) {
            stats.min = value;
        }
        if (stats.sampleCount == 0 || value > stats.max) {
            stats.max = value;
        }
        
        // 更新平均值（增量计算）
        float oldAvg = stats.avg;
        stats.sampleCount++;
        stats.avg = oldAvg + (value - oldAvg) / stats.sampleCount;
        
        // 更新 RMS（均方根）
        stats.rms = std::sqrt(stats.rms * stats.rms * (stats.sampleCount - 1) / stats.sampleCount +
                             value * value / stats.sampleCount);
    }
    
    emit statsUpdated(deviceId);
}

ChannelStats DataProcessor::calculateStats(const QVector<float>& values)
{
    ChannelStats stats;
    
    if (values.isEmpty()) {
        return stats;
    }
    
    stats.min = values[0];
    stats.max = values[0];
    float sum = 0;
    float sumSq = 0;
    
    for (float v : values) {
        if (v < stats.min) stats.min = v;
        if (v > stats.max) stats.max = v;
        sum += v;
        sumSq += v * v;
    }
    
    stats.sampleCount = values.size();
    stats.avg = sum / stats.sampleCount;
    stats.rms = std::sqrt(sumSq / stats.sampleCount);
    
    return stats;
}

} // namespace HeteroLink

File: software/host/src/core/data_processor.rs
use crate::domain::{TelemetryData, ChannelStats};
use std::collections::{HashMap, VecDeque};
use std::time::{SystemTime, UNIX_EPOCH};
use tokio::sync::RwLock;
use tracing::info;

/// 设备数据（内部结构）
struct DeviceData {
    buffer: VecDeque<TelemetryData>,
    channel_stats: HashMap<i32, ChannelStats>,
    last_update: u64,
    max_buffer_size: usize,
}

impl DeviceData {
    fn new(buffer_size: usize) -> Self {
        Self {
            buffer: VecDeque::new(),
            channel_stats: HashMap::new(),
            last_update: 0,
            max_buffer_size: buffer_size,
        }
    }
}

/// 数据处理器
///
/// 使用 `RwLock` 替代 `Mutex`，允许并发读取（高频统计查询场景下更优）。
/// 外层已由 `Arc<DataProcessor>` 共享，内部无需再套 Arc。
pub struct DataProcessor {
    device_data: RwLock<HashMap<String, DeviceData>>,
    default_buffer_size: usize,
}

impl DataProcessor {
    pub fn new() -> Self {
        Self {
            device_data: RwLock::new(HashMap::new()),
            default_buffer_size: 10000,
        }
    }

    pub async fn set_buffer_size(&self, device_id: &str, size: usize) {
        let mut device_data = self.device_data.write().await;
        if let Some(data) = device_data.get_mut(device_id) {
            data.max_buffer_size = size;
            while data.buffer.len() > size {
                data.buffer.pop_front();
            }
        } else {
            device_data.insert(device_id.to_string(), DeviceData::new(size));
        }
        info!("Buffer size set for device {}: {}", device_id, size);
    }

    pub async fn add_data(&self, device_id: &str, data: TelemetryData) {
        let mut device_data = self.device_data.write().await;
        let dev_data = device_data
            .entry(device_id.to_string())
            .or_insert_with(|| DeviceData::new(self.default_buffer_size));

        dev_data.buffer.push_back(data.clone());
        while dev_data.buffer.len() > dev_data.max_buffer_size {
            dev_data.buffer.pop_front();
        }

        dev_data.last_update = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        Self::update_stats(dev_data, &data);
    }

    pub async fn get_data(&self, device_id: &str) -> Vec<TelemetryData> {
        let device_data = self.device_data.read().await;
        device_data
            .get(device_id)
            .map(|d| d.buffer.iter().cloned().collect())
            .unwrap_or_default()
    }

    pub async fn get_latest_data(&self, device_id: &str, count: usize) -> Vec<TelemetryData> {
        let device_data = self.device_data.read().await;
        device_data
            .get(device_id)
            .map(|d| {
                let start = d.buffer.len().saturating_sub(count);
                d.buffer.iter().skip(start).cloned().collect()
            })
            .unwrap_or_default()
    }

    pub async fn get_stats(&self, device_id: &str) -> HashMap<i32, ChannelStats> {
        let device_data = self.device_data.read().await;
        device_data
            .get(device_id)
            .map(|d| d.channel_stats.clone())
            .unwrap_or_default()
    }

    pub async fn clear_data(&self, device_id: &str) {
        let mut device_data = self.device_data.write().await;
        if let Some(data) = device_data.get_mut(device_id) {
            data.buffer.clear();
            data.channel_stats.clear();
        }
    }

    pub async fn clear_all(&self) {
        self.device_data.write().await.clear();
        info!("All data cleared");
    }

    fn update_stats(dev_data: &mut DeviceData, data: &TelemetryData) {
        for (i, &value) in data.channels.iter().enumerate() {
            let i = i as i32;
            let stats = dev_data
                .channel_stats
                .entry(i)
                .or_default();

            if stats.sample_count == 0 || value < stats.min {
                stats.min = value;
            }
            if stats.sample_count == 0 || value > stats.max {
                stats.max = value;
            }

            let old_avg = stats.avg;
            stats.sample_count += 1;
            stats.avg = old_avg + (value - old_avg) / stats.sample_count as f32;

            let count = stats.sample_count as f32;
            stats.rms =
                (stats.rms * stats.rms * (count - 1.0) / count + value * value / count).sqrt();
        }
    }
}
